### ISSUE TITLE
Some sudden peaks in generated results.

### ISSUE THREAD

**Issue by chaiyujin** [G-Wang/WaveRNN-Pytorch#7]
Hi, it's a nice work to make the fatchord's model easier to use. I trained this model with 'bits' mode in my own framework, that is to say my pre-processing and lr schedule are different. I also successfully trained a model, quite easier to train than the original wavernn.

However, my generated results with mels from tacotron have some sudden peaks. As you can see in the following picture, there are two abnormal peaks (one reaches 1.0 and another reaches -1.0).

![pred_12_spkid_1](https://user-images.githubusercontent.com/10378898/49848218-11b77380-fe0f-11e8-9eac-af7548d40aa2.png)

Have you ever encountered this problem with 'bits' mode? It also appear in my trained softmax-wavenet. I wonder to figure out if it is caused by my (maybe wrong) pre-processing or by discrezited output.

**Comment by G-Wang** [comment]
hmm, I don't recall seeing these artifiacts using the "bits" mode, I do see them when I'm training on "raw" with gaussian, but that's cause my guassian output is unbounded so when I clip I get these 1.0 or -1.0 wav peaks.

I will have to look at some generated audio from my bit training results to see if i can find these.

Have you tried comparing your tacotron Mel output on a training sentence with the ground truth Mel for the same sentence and see if there's any difference in the mel spectrogram? Also are you training the wavernn on tacotron generated mel spectrogram or just the ground-truth real audio mel spectrogram?

The only thing I can think of is that there may be slight differences in the audio preprocessing step between your tacotron training and the wavernn training, perhaps in preemphasis or normalization? Note that my audio preprocessing I use `lws` to generate spectrograms (so I can use r9r9's text-mel frontends), where you may have used `librosa` for generating spectrograms, so there may be differences here.

One thing you could do is to copy the `audio.py`, `hparams.py` from your tacotron training script and call them in my `preprocess.py` to generate mel training data exactly same as your tacotron, then train a new 9-bit wavernn on the data.

You would just have to change this audio import here: https://github.com/G-Wang/WaveRNN-Pytorch/blob/b3911f7092ae4add610030da52a67f24a4ce7d6d/preprocess.py#L14 to `from my_audio import *` and ensure that there is a `melspectrogram (wav)` function in there, similar to what's in my `audio.py` here:
https://github.com/G-Wang/WaveRNN-Pytorch/blob/b3911f7092ae4add610030da52a67f24a4ce7d6d/audio.py#L46

**Comment by chaiyujin** [comment]
Thanks for your detailed explanation. Since you didn't find any artifacts, it should be my fault on pre-processing. I'll check that.

About your questions:
- I trained with ground-truth mel. Even though generate audio with ground-truth mel, there're these artifacts as well.
- I use same pre-processed data pairs for tacotron and wavernn.
- However, I implement pre-processing by myself using a torch.nn.Module to tansform mel spectrograms.    I compared my processed data with `lws`. They looks similar but not totally equal.
- I found that you didn't rescale audio in preprocessing, `rescaling=False`. I keep it as `True`, because r9y9 set it as true in his wavenet repo.

**Comment by G-Wang** [comment]
Yes, I was using r9r9's deepvoice 3 nyanko build (https://github.com/r9y9/deepvoice3_pytorch/blob/master/presets/nyanko_ljspeech.json) for text to Mel front end,which is pretty much dctts. I just copied that preset hparams over to my WaveRNN repo.

**Comment by geneing** [comment]
@chaiyujin I'm seeing spikes too. I'm using 10 bit raw, librosa library. The spikes do become more rare after longer training, but I think there is some subtle problem.
